test/e2e/network/network_policy.go

```diff
@@ -821,49 +821,39 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
 			const allowedPort = 80
-			ginkgo.By("Creating a network policy for the server which allows traffic from client-a-updated.")
+			ginkgo.By("Creating a network policy for the server which denies all traffic.")
 			policy := &networkingv1.NetworkPolicy{
 				ObjectMeta: metav1.ObjectMeta{
-					Name: "allow-pod-b-via-pod-selector",
+					Name: "deny-ingress-via-isolated-label-selector",
 				},
 				Spec: networkingv1.NetworkPolicySpec{
 					PodSelector: metav1.LabelSelector{
 						MatchLabels: map[string]string{
 							"pod-name": podServerLabelSelector,
 						},
-					},
-					Ingress: []networkingv1.NetworkPolicyIngressRule{{
-						From: []networkingv1.NetworkPolicyPeer{{
-							PodSelector: &metav1.LabelSelector{
-								MatchExpressions: []metav1.LabelSelectorRequirement{{
-									Key:      "pod-name",
-									Operator: metav1.LabelSelectorOpDoesNotExist,
-								}},
-							},
+						MatchExpressions: []metav1.LabelSelectorRequirement{{
+							Key:      "isolated",
+							Operator: metav1.LabelSelectorOpExists,
 						}},
-					}},
+					},
+					PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+					Ingress:     []networkingv1.NetworkPolicyIngressRule{},
 				},
 			}
 
 			policy, err := f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
 			framework.ExpectNoError(err, "Error creating Network Policy %v: %v", policy.ObjectMeta.Name, err)
 			defer cleanupNetworkPolicy(f, policy)
 
-			ginkgo.By(fmt.Sprintf("Creating client pod %s that should not be able to connect to %s.", "client-a", service.Name))
-			podClient := createNetworkClientPod(f, f.Namespace, "client-a", service, allowedPort)
-			defer func() {
-				ginkgo.By(fmt.Sprintf("Cleaning up the pod %s", podClient.Name))
-				if err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete(podClient.Name, nil); err != nil {
-					framework.Failf("unable to cleanup pod %v: %v", podClient.Name, err)
-				}
-			}()
-			checkNoConnectivity(f, f.Namespace, podClient, service)
-
-			ginkgo.By(fmt.Sprintf("Updating client pod %s that should successfully connect to %s.", podClient.Name, service.Name))
-			podClient = updateNetworkClientPodLabel(f, f.Namespace, podClient.Name, "replace", "/metadata/labels", map[string]string{})
-			checkConnectivity(f, f.Namespace, podClient, service)
+			// Client can connect to service when the server pod is not selected by the network policy.
+			testCanConnect(f, f.Namespace, "client-a", service, allowedPort)
+
+			// Client cannot connect to service after updating the server pod's labels to match the network policy.
+			ginkgo.By(fmt.Sprintf("Updating server pod %s to be selected by network policy %s.", podServer.Name, policy.Name))
+			updateNetworkClientPodLabel(f, f.Namespace, podServer.Name, "add", "/metadata/labels", map[string]string{"isolated": ""})
+			testCannotConnect(f, f.Namespace, "client-a", service, allowedPort)
```

Comment:
Pod client has `RestartPolicyNever`, so would this also fail since pod will be in `Succeeded` state from previous test case?

Reply:
`testCanConnect` and `testCannotConnect` create new client Pod, check connection, and clean up it. 
https://github.com/kubernetes/kubernetes/blob/52d7614a8ca5b8aebc45333b6dc8fbf86a5e7ddf/test/e2e/network/network_policy.go#L1386-L1396
Using same pod name `client-a` is to emphasize the behavior changes by server pod's label change only. The two steps actually use two client Pods having same name and labels.

Reply:
So this is no longer "ingress access to updated pod", but "ingress access to new pod" then? I think we still want the updated pod test case.

Reply:
It's still updated pod, but just updated server Pod.
As the PR description explains, all network policy tests are using the utils "testCannotConnect" and "testCanConnect" which creates a client Pod designed for **one-shot** check without restarting, it's not convenient to test twice with a single client Pod (that's also the mistake the original test made), unless we write another util just for this case.
I think essentially the updated pod test case wants to ensure network policy can be enabled on existing Pods after updating their labels to match them, so it should make sense to verify it on server Pod, as even the original test just verified it on client Pod. Does it make sense?

Reply:
I see, so if we are creating new client pod every time and only the server pod gets updated labels, it seems like we can still test the `deny` -> `allow` test case right and we don't really need to flip the order? I think we could benefit from having both:
* deny -> allow by updating server pod labels with new client pod for each connection
* allow -> deny by updating server pod labels with new client pod for each connection

Reply:
I updated the order because the server Pod and its service is created in `BeforeEach`, having only a fixed label `pod-name: server` which is also used by its service to select backends. So it's not convenient to first create a deny policy which selects the server Pod, and then update the server Pod to not match it as updating label "pod-name: server" will make service not work.
To have both tests, I can append one more step to remove the "isolated" label and call `testCanConnect` again, it would be the reversed version of your proposal, what do you think?

Reply:
This looks like a different test case to me. The original one is to change the network policy ingress rule selector, so a client that couldn't connect can connect now. Your test case is to change network policy selector, so a server that was protected by the policy become not protected. Maybe we can fix the original one by creating a new client pod and create a new test case for this change? Creating a new client pod is not ideal to test policy change but the enforcement is on the server side for both cases anyway.

Reply:
Thanks @lzang for reviewing. I restored the original test and fixed it by creating the client pod with `OnFailure` RestartPolicy and updating its label after failure is detected as creating a new client pod may make the test case same as some existing tests which check pods with different labels will have different behaviors, like 
https://github.com/kubernetes/kubernetes/blob/29b09c7fb1a8d38ab9df4873553cc0b8d97aae95/test/e2e/network/network_policy.go#L147-L152
I created a new test case to ensure network policy selector can take effect on updated server pod as you suggested.

Reply:
The existing one doesn't have policy update, so I don't think they are the same, but I am fine if the new approach by modifying the restart policy could work.